src/feature/hs/hs_intropoint.c

```diff
@@ -181,6 +181,108 @@ hs_intro_send_intro_established_cell,(or_circuit_t *circ))
   return ret;
 }
 
+static void
+handle_establish_intro_cell_dos_extension(
+                                const trn_cell_extension_field_t *field,
+                                or_circuit_t *circ)
+{
+  ssize_t ret;
+  uint64_t intro2_rate_per_sec = 0, intro2_burst_per_sec = 0;
+  trn_cell_extension_dos_t *dos = NULL;
+
+  tor_assert(field);
+  tor_assert(circ);
+
+  ret = trn_cell_extension_dos_parse(&dos,
+                 trn_cell_extension_field_getconstarray_field(field),
+                 trn_cell_extension_field_getlen_field(field));
+  if (ret < 0) {
+    goto end;
+  }
+
+  for (size_t i = 0; i < trn_cell_extension_dos_get_n_params(dos); i++) {
+    const trn_cell_extension_dos_param_t *param =
+      trn_cell_extension_dos_getconst_params(dos, i);
+    if (BUG(param == NULL)) {
+      goto end;
+    }
+
+    switch (trn_cell_extension_dos_param_get_type(param)) {
+    case TRUNNEL_DOS_PARAM_TYPE_INTRO2_RATE_PER_SEC:
+      intro2_rate_per_sec = trn_cell_extension_dos_param_get_value(param);
+      break;
+    case TRUNNEL_DOS_PARAM_TYPE_INTRO2_BURST_PER_SEC:
+      intro2_burst_per_sec = trn_cell_extension_dos_param_get_value(param);
+      break;
+    default:
+      goto end;
+    }
+  }
+
+  /* Validation. A value of 0 on either of them means the defenses are
+   * disabled so we ignore. */
+  if ((intro2_rate_per_sec > INT32_MAX || intro2_rate_per_sec == 0) ||
+      (intro2_burst_per_sec > INT32_MAX || intro2_burst_per_sec == 0) ||
+      (intro2_burst_per_sec < intro2_rate_per_sec)) {
+    circ->introduce2_dos_defense_enabled = 0;
+    log_info(LD_REND, "Intro point DoS defenses disabled due to bad values");
+  } else {
+    circ->introduce2_dos_defense_enabled = 1;
+
+    /* Initialize the INTRODUCE2 token bucket for the rate limiting. */
+    token_bucket_ctr_init(&circ->introduce2_bucket,
+                          (uint32_t) intro2_rate_per_sec,
+                          (uint32_t) intro2_burst_per_sec,
+                          (uint32_t) approx_time());
+    log_debug(LD_REND, "Intro point DoS defenses enabled. Rate is %" PRIu64
+                       " and Burst is %" PRIu64, intro2_rate_per_sec,
+                       intro2_burst_per_sec);
+  }
+
+ end:
+  trn_cell_extension_dos_free(dos);
+  return;
+}
+
+static void
+handle_establish_intro_cell_extensions(
+                            const trn_cell_establish_intro_t *parsed_cell,
+                            or_circuit_t *circ)
+{
+  const trn_cell_extension_t *extensions;
+
+  tor_assert(parsed_cell);
+  tor_assert(circ);
+
+  extensions = trn_cell_establish_intro_getconst_extensions(parsed_cell);
+  if (extensions == NULL) {
+    goto end;
+  }
+
+  /* Go over all extensions. */
```

Comment:
You mentioned that this `N_EXTENSIONS` `N_PARAMS` inception situation is confusing and I can see why this is that. Since we basically have:
```
     AUTH_KEY_TYPE         [1 byte]
     AUTH_KEY_LEN          [2 bytes]
     AUTH_KEY              [AUTH_KEY_LEN bytes]
     N_EXTENSIONS          [1 byte]
     N_EXTENSIONS times:
        EXT_FIELD_TYPE     [1 byte]
        EXT_FIELD_LEN      [1 byte]
        N_PARAMS    [1 byte]
        N_PARAMS times:
           PARAM_TYPE  [1 byte]
           PARAM_VALUE [8 byte]
     HANDSHAKE_AUTH        [MAC_LEN bytes]
     SIG_LEN               [2 bytes]
     SIG                   [SIG_LEN bytes]
```

Just for the sake of argument and sorry for suggesting it late in the process, but what would happen if we ditched the second layer of abstraction and encoded rate and burst directly as `N_EXTENSIONS` instead of `N_PARAMS`?

Reply:
Not sure we can. The `N_EXTENSIONS` contains an extension itself. In this case, it contains the "INTRO2 DoS Extension" encoded data.

Which in turn has this `N_PARAMS` design to list the parameters. There could be other extensions in the future that have nothing to do with this DoS one.

Reply:
Hm, let me give you a visual. Here is what prop305 is doing:
```
     AUTH_KEY_TYPE                          [1 byte]
     AUTH_KEY_LEN                           [2 bytes]
     AUTH_KEY                               [AUTH_KEY_LEN bytes]
     N_EXTENSIONS (1)                       [1 byte]
        EXT_FIELD_TYPE (DOS_PARAMETERS)     [1 byte]
        EXT_FIELD_LEN                       [1 byte]
        N_PARAMS (2)                        [1 byte]
        N_PARAMS times:
           PARAM_TYPE (DOS_INTRODUCE2_RATE_PER_SEC)    [1 byte]
           PARAM_VALUE (25)                      [8 byte]
           PARAM_TYPE (DOS_INTRODUCE2_BURST_PER_SEC)    [1 byte]
           PARAM_VALUE (250)                      [8 byte]
     HANDSHAKE_AUTH        [MAC_LEN bytes]
     SIG_LEN               [2 bytes]
     SIG                   [SIG_LEN bytes]
```

I'm wondering if we can do:
```
     AUTH_KEY_TYPE                          [1 byte]
     AUTH_KEY_LEN                           [2 bytes]
     AUTH_KEY                               [AUTH_KEY_LEN bytes]
     N_EXTENSIONS (2)                       [1 byte]
        EXT_FIELD_TYPE (DOS_INTRODUCE2_RATE_PER_SEC)    [1 byte]
        EXT_FIELD_VALUE (25)                      [8 byte]
        EXT_FIELD_TYPE (DOS_INTRODUCE2_BURST_PER_SEC)    [1 byte]
        EXT_FIELD_VALUE (250)                      [8 byte]
     HANDSHAKE_AUTH        [MAC_LEN bytes]
     SIG_LEN               [2 bytes]
     SIG                   [SIG_LEN bytes]
```

Reply:
Right. No we can not :).

Because the `EXT_FIELD_TYPE` is describing an *extension* type, can't be used to host the actual extension data.

Furthermore, we need to specify the `length` of the extension which is put in the `EXT_FIELD_LEN`. And also, considering that a ESTABLISH_INTRO cell could have many more different extensions, we need a way to specify which extension you are looking at when iterating over all extensions.

In our case, we are adding 1 new extension that is the "DoS defenses" so `N_EXTENSIONS` has to be 1.

This extension concept is spec out in all our HS cells so we can't move away from it.

Reply:
After discussion on IRC, this seems like a possible thing, but we should not do it because it breaks the semantics of extensions.